File: crmessenger-telegram/internal/repo/amocrm/repo.py
from opentelemetry.trace import Status, StatusCode, SpanKind

from .sql_query import *
from internal import interface
from internal import model


class AmocrmRepo(interface.IAmocrmRepo):
    def __init__(self, db: interface.IDB, tel: interface.ITelemetry):
        self.db = db
        self.tracer = tel.tracer()
        self.logger = tel.logger()

    async def create_amocrm_source(
            self,
            userbot_tg_user_id: int,
            amocrm_source_id: int,
            amocrm_pipeline_id: int,
            amocrm_external_id: str,
            amocrm_scope_id: str,
    ) -> int:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.create_amocrm_source",
                kind=SpanKind.INTERNAL,
                attributes={
                    "amocrm_source_id": amocrm_source_id,
                    "amocrm_pipeline_id": amocrm_pipeline_id,
                    "amocrm_external_id": amocrm_external_id,
                    "amocrm_scope_id": amocrm_scope_id,
                    "userbot_tg_user_id": userbot_tg_user_id,
                }
        ) as span:
            try:
                args = {
                    "userbot_tg_user_id": userbot_tg_user_id,
                    "amocrm_source_id": amocrm_source_id,
                    "amocrm_pipeline_id": amocrm_pipeline_id,
                    "amocrm_external_id": amocrm_external_id,
                    "amocrm_scope_id": amocrm_scope_id,
                }
                amocrm_id = await self.db.insert(create_amocrm_source, args)

                span.set_status(StatusCode.OK)
                return amocrm_id
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def amocrm_source_by_userbot_tg_user_id(self, userbot_tg_user_id: int) -> list[model.AmocrmSource]:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.amocrm_source_by_userbot_tg_user_id",
                kind=SpanKind.INTERNAL,
                attributes={
                    "userbot_tg_user_id": userbot_tg_user_id,
                }
        ) as span:
            try:
                args = {
                    "userbot_tg_user_id": userbot_tg_user_id,
                }
                rows = await self.db.select(amocrm_source_by_userbot_tg_user_id, args)
                if rows:
                    rows = model.AmocrmSource.serialize(rows)

                span.set_status(StatusCode.OK)
                return rows
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def create_chat(self, contact_id: int, amocrm_conversation_id: str, amocrm_chat_id: str):
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.create_chat",
                kind=SpanKind.INTERNAL,
                attributes={
                    "contact_id": contact_id,
                    "amocrm_conversation_id": amocrm_conversation_id,
                    "amocrm_chat_id": amocrm_chat_id
                }
        ) as span:
            try:
                args = {
                    'contact_id': contact_id,
                    'amocrm_conversation_id': amocrm_conversation_id,
                    'amocrm_chat_id': amocrm_chat_id
                }
                chat_id = await self.db.insert(create_chat, args)

                span.set_status(StatusCode.OK)
                return chat_id
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def amocrm_source_by_amocrm_external_id(self, amocrm_external_id: str) -> list[model.AmocrmSource]:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.amocrm_source_by_amocrm_external_id",
                kind=SpanKind.INTERNAL,
                attributes={
                    "amocrm_external_id": amocrm_external_id,
                }
        ) as span:
            try:
                args = {
                    "amocrm_external_id": amocrm_external_id,
                }
                rows = await self.db.select(amocrm_source_by_amocrm_external_id, args)
                if rows:
                    rows = model.AmocrmSource.serialize(rows)

                span.set_status(StatusCode.OK)
                return rows
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def chat_by_contact_id(self, contact_id: int) -> list[model.Chat]:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.chat_by_contact_id",
                kind=SpanKind.INTERNAL,
                attributes={
                    "contact_id": contact_id,
                }
        ) as span:
            try:
                args = {'contact_id': contact_id}
                rows = await self.db.select(chat_by_contact_id, args)
                if rows:
                    rows = model.Chat.serialize(rows)

                span.set_status(StatusCode.OK)
                return rows
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def chat_by_amocrm_chat_id(self, amocrm_chat_id: int) -> list[model.Chat]:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.chat_by_amocrm_chat_id",
                kind=SpanKind.INTERNAL,
                attributes={
                    "amocrm_chat_id": amocrm_chat_id,
                }
        ) as span:
            try:
                args = {'amocrm_chat_id': amocrm_chat_id}
                rows = await self.db.select(chat_by_amocrm_chat_id, args)
                if rows:
                    rows = model.Chat.serialize(rows)

                span.set_status(StatusCode.OK)
                return rows
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def create_contact(
            self,
            first_name: str,
            last_name: str,
            tg_chat_id: int,
            amocrm_contact_id: int,
    ) -> int:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.create_contact",
                kind=SpanKind.INTERNAL,
                attributes={
                    "first_name": first_name,
                    "last_name": last_name,
                    "tg_chat_id": tg_chat_id,
                    "amocrm_contact_id": amocrm_contact_id,
                }
        ) as span:
            try:
                args = {
                    'first_name': first_name,
                    'last_name': last_name,
                    'tg_chat_id': tg_chat_id,
                    'amocrm_contact_id': amocrm_contact_id
                }
                contact_id = await self.db.insert(create_contact, args)

                span.set_status(StatusCode.OK)
                return contact_id
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def contact_by_tg_chat_id(self, tg_chat_id: int) -> list[model.Contact]:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.contact_by_tg_chat_id",
                kind=SpanKind.INTERNAL,
                attributes={
                    "tg_chat_id": tg_chat_id,
                }
        ) as span:
            try:
                args = {'tg_chat_id': tg_chat_id}
                rows = await self.db.select(contact_by_tg_chat_id, args)
                if rows:
                    rows = model.Contact.serialize(rows)

                span.set_status(StatusCode.OK)
                return rows
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def contact_by_id(self, contact_id: int) -> list[model.Contact]:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.contact_by_id",
                kind=SpanKind.INTERNAL,
                attributes={
                    "contact_id": contact_id,
                }
        ) as span:
            try:
                args = {'contact_id': contact_id}
                rows = await self.db.select(contact_by_id, args)
                if rows:
                    rows = model.Contact.serialize(rows)

                span.set_status(StatusCode.OK)
                return rows
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def create_message(self, chat_id: int, amocrm_message_id: str, role: str, text: str) -> int:
        with self.tracer.start_as_current_span(
                "AmocrmSourceRepo.create_message",
                kind=SpanKind.INTERNAL,
                attributes={
                    "chat_id": chat_id,
                    "amocrm_message_id": amocrm_message_id,
                    "role": role,
                    "text": text
                }
        ) as span:
            try:
                args = {
                    "chat_id": chat_id,
                    "amocrm_message_id": amocrm_message_id,
                    "role": role,
                    "text": text
                }
                message_id = await self.db.insert(create_message, args)

                span.set_status(StatusCode.OK)
                return message_id
            except Exception as e:
                span.record_exception(e)
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

File: crmessenger-whatsapp/internal/repo/amocrm/repo.go
package amocrm

import (
	"context"
	"crmessenger-whatsapp/internal/model"
	"fmt"
	"github.com/georgysavva/scany/v2/pgxscan"
	"github.com/jackc/pgx/v5"
)

func New(db model.IDB) *AmocrmSourceRepo {
	return &AmocrmSourceRepo{db: db}
}

type AmocrmSourceRepo struct {
	db model.IDB
}

func (r *AmocrmSourceRepo) CreateAmocrmSource(ctx context.Context,
	amocrmSourceID,
	amocrmPipelineID int,
	userbotWaPhoneNumber,
	amocrmExternalID,
	amocrmScopeID string,
) error {
	args := pgx.NamedArgs{
		"userbot_wa_phone_number": userbotWaPhoneNumber,
		"amocrm_source_id":        amocrmSourceID,
		"amocrm_pipeline_id":      amocrmPipelineID,
		"amocrm_external_id":      amocrmExternalID,
		"amocrm_scope_id":         amocrmScopeID,
	}
	_, err := r.db.Insert(ctx, CreateAmocrmSource, args)
	if err != nil {
		return err
	}
	return nil
}

func (r *AmocrmSourceRepo) AmocrmSourceByUserbotWaPhoneNumber(ctx context.Context,
	userbotWaPhoneNumber string,
) ([]*model.AmocrmSource, error) {
	args := pgx.NamedArgs{
		"userbot_wa_phone_number": userbotWaPhoneNumber,
	}
	rows, err := r.db.Select(ctx, AmocrmSourceByWaPhoneNumber, args)
	if err != nil {
		return nil, err
	}

	var source []*model.AmocrmSource
	err = pgxscan.ScanAll(&source, rows)
	if err != nil {
		return nil, err
	}

	return source, nil
}

func (r *AmocrmSourceRepo) AmocrmSourceByAmocrmExternalID(ctx context.Context,
	amocrmExternalID string,
) ([]*model.AmocrmSource, error) {
	args := pgx.NamedArgs{
		"amocrm_external_id": amocrmExternalID,
	}
	rows, err := r.db.Select(ctx, AmocrmSourceByExternalID, args)
	if err != nil {
		return nil, err
	}

	var source []*model.AmocrmSource
	err = pgxscan.ScanAll(&source, rows)
	if err != nil {
		return nil, err
	}

	return source, nil
}

func (r *AmocrmSourceRepo) CreateContact(ctx context.Context,
	contactWaPhoneNumber string,
	amocrmContactID int,
) (int, error) {
	args := pgx.NamedArgs{
		"wa_phone_number":   contactWaPhoneNumber,
		"amocrm_contact_id": amocrmContactID,
	}
	contactID, err := r.db.Insert(ctx, CreateContact, args)
	if err != nil {
		return 0, err
	}

	switch contactID := contactID.(type) {
	case int:
		return contactID, nil
	case int64:
		return int(contactID), nil
	case int32:
		return int(contactID), nil
	}
	return 0, fmt.Errorf("type %T is not supported", contactID)
}

func (r *AmocrmSourceRepo) ContactByWaPhoneNumber(ctx context.Context,
	contactWaPhoneNumber string,
) ([]*model.AmocrmContact, error) {
	args := pgx.NamedArgs{
		"wa_phone_number": contactWaPhoneNumber,
	}
	rows, err := r.db.Select(ctx, ContactByWaPhoneNumber, args)
	if err != nil {
		return nil, err
	}

	var contact []*model.AmocrmContact
	err = pgxscan.ScanAll(&contact, rows)
	if err != nil {
		return nil, err
	}

	return contact, nil
}

func (r *AmocrmSourceRepo) ContactByID(ctx context.Context,
	contactID int,
) ([]*model.AmocrmContact, error) {
	args := pgx.NamedArgs{
		"contact_id": contactID,
	}
	rows, err := r.db.Select(ctx, ContactByID, args)
	if err != nil {
		return nil, err
	}

	var contact []*model.AmocrmContact
	err = pgxscan.ScanAll(&contact, rows)
	if err != nil {
		return nil, err
	}

	return contact, nil
}

func (r *AmocrmSourceRepo) CreateChat(ctx context.Context,
	contactID int,
	amocrmConversationID,
	amocrmChatID string,
) (int, error) {
	args := pgx.NamedArgs{
		"contact_id":             contactID,
		"amocrm_conversation_id": amocrmConversationID,
		"amocrm_chat_id":         amocrmChatID,
	}
	fmt.Println(args)
	chatID, err := r.db.Insert(ctx, CreateChat, args)
	if err != nil {
		return 0, err
	}
	switch chatID := chatID.(type) {
	case int:
		return chatID, nil
	case int64:
		return int(chatID), nil
	case int32:
		return int(chatID), nil
	}
	return 0, fmt.Errorf("type %T is not supported", contactID)
}

func (r *AmocrmSourceRepo) ChatByContactID(ctx context.Context,
	contactID int,
) ([]*model.AmocrmChat, error) {
	args := pgx.NamedArgs{
		"contact_id": contactID,
	}
	rows, err := r.db.Select(ctx, ChatByContactID, args)
	if err != nil {
		return nil, err
	}

	var chat []*model.AmocrmChat
	err = pgxscan.ScanAll(&chat, rows)
	if err != nil {
		return nil, err
	}

	return chat, nil
}

func (r *AmocrmSourceRepo) ChatByAmocrmChatID(ctx context.Context,
	amocrmChatID string,
) ([]*model.AmocrmChat, error) {
	args := pgx.NamedArgs{
		"amocrm_chat_id": amocrmChatID,
	}
	rows, err := r.db.Select(ctx, ChatByAmocrmChatID, args)
	if err != nil {
		return nil, err
	}

	var chat []*model.AmocrmChat
	err = pgxscan.ScanAll(&chat, rows)
	if err != nil {
		return nil, err
	}

	return chat, nil
}

func (r *AmocrmSourceRepo) CreateMessage(ctx context.Context,
	chatID int,
	amocrmMessageID,
	role,
	text string,
) error {
	args := pgx.NamedArgs{
		"chat_id":           chatID,
		"amocrm_message_id": amocrmMessageID,
		"role":              role,
		"text":              text,
	}
	_, err := r.db.Insert(ctx, CreateMessage, args)
	if err != nil {
		return err
	}
	return nil
}
